codeworld-server/src/Util.hs

```diff
@@ -296,3 +299,53 @@ removeDirectoryIfExists dirName =
     handleExists e
         | isDoesNotExistError e = return ()
         | otherwise = throwIO e
+
+loadDumpedTree :: BuildMode -> UserId -> IO DirTree
+loadDumpedTree bm uid = do
+    let file = userProjectDir bm uid </> "tree.info"
+    exists <- doesFileExist file
+    realTree <- getDirectoryTree bm uid
+    case exists of
+        False -> return realTree
+        True -> do
+            content  <- LB.readFile file
+            return $ case decode content of
+                Nothing -> realTree
+                Just dumpedTree -> case compareTrees realTree (Dir "root" dumpedTree) of
+                    True  -> Dir "root" dumpedTree
+                    False -> realTree
+
+getDirectoryTree :: BuildMode -> UserId -> IO DirTree
+getDirectoryTree bm uid = do
+    ftr <- getDirectory $ userProjectDir bm uid
+    let (Node rootPath children) = toTree ftr
+    children' <- mapM (constructTree rootPath) $ filter notInfo children
+    return $ Dir "root" children'
+    where
+        constructTree :: FilePath -> Tree FilePath -> IO DirTree
+        constructTree prefix (Node path [(Node "dir.info" [])]) = do -- empty directory
+            name <- B.readFile $ prefix </> path </> "dir.info"
+            return $ Dir (T.decodeUtf8 name) []
+        constructTree prefix (Node path []) = do
+            let currentNode = prefix </> path
+            source <- LB.readFile currentNode
```

Comment:
So you're going to read every single user project into memory just to read a directory?  That seems excessive.  Why not just remember the directory structures, and the project names, so you can load them when they are needed later?  There is already a `loadProject` endpoint that does that.

Reply:
Okay, lazy loading is point I missed here.

In this solution user make one directory tree request per session and copy whole structure to js. After that changes to tree (moving, deleting, creating of nodes etc) becomes after 200 response from server, browser does not re ask all tree from server until next session.

I need get name of projects, but I can not do it without loading and decoding. Storing all tree in browser makes search pure js task without server requests.

How much entries in biggest projects directory in production?

Reply:
You're right, you definitely need to load the files in order to determine their names.  That's unfortunate.  But I think we should decide not to keep all of these files in memory, in any case.  Each file can be up to a megabyte or so (and that limit is a bug), and there could easily be a hundred projects in someone's directory tree.  There probably are in mine.

Reply:
"There probably are in mine."

I implemented lazy loading for this tree in last commit.

Now I going to implement route with updating of metadata, to set order of files in directory.